core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala

```diff
@@ -199,41 +222,131 @@ private[spark] class HeartbeatReceiver(sc: SparkContext, clock: Clock)
-            // 2) call scheduler.executorLost() underlying to fail any tasks assigned to
-            //    those executors to avoid app hang
-            sc.schedulerBackend match {
-              case backend: CoarseGrainedSchedulerBackend =>
-                backend.driverEndpoint.send(RemoveExecutor(executorId,
-                  ExecutorProcessLost(
-                    s"Executor heartbeat timed out after ${now - lastSeenMs} ms")))
-
-              // LocalSchedulerBackend is used locally and only has one single executor
-              case _: LocalSchedulerBackend =>
-
-              case other => throw new UnsupportedOperationException(
-                s"Unknown scheduler backend: ${other.getClass}")
-            }
+    if (!sc.conf.get(HEARTBEAT_RECEIVER_CHECK_WORKER_LAST_HEARTBEAT)) {
+      for ((executorId, lastSeenMs) <- executorLastSeen) {
+        if (now - lastSeenMs > executorTimeoutMs) {
+          killExecutor(executorId, now - lastSeenMs)
+          waitingList.remove(executorId)
+          executorLastSeen.remove(executorId)
+        }
+      }
+    } else {
+      for ((executorId, workerLastHeartbeat) <- waitingList) {
+        if (now - workerLastHeartbeat > executorTimeoutMs / 2) {
+          killExecutor(executorId, now - workerLastHeartbeat)
+          waitingList.remove(executorId)
+          executorLastSeen.remove(executorId)
+        }
+      }
+
+      val buf = new ArrayBuffer[String]()
+      for ((executorId, lastSeenMs) <- executorLastSeen) {
+        if (now - lastSeenMs > executorTimeoutMs) {
+          sc.schedulerBackend match {
+            case _: StandaloneSchedulerBackend =>
+              buf += executorId
+            case _ =>
+              killExecutor(executorId, now - lastSeenMs)
+              waitingList.remove(executorId)
+              executorLastSeen.remove(executorId)
+          }
+        }
+      }
+
+      sc.schedulerBackend match {
+        case backend: StandaloneSchedulerBackend =>
+          backend.client.workerLastHeartbeat(sc.applicationId, buf) match {
+            case Some(workerLastHeartbeats) =>
+              for ((executorId, workerLastHeartbeat) <- buf zip workerLastHeartbeats) {
+                if (now - workerLastHeartbeat > executorTimeoutMs / 2) {
+                  val lastSeenMs = executorLastSeen.get(executorId).get
+                  killExecutor(executorId, now - lastSeenMs)
+                  waitingList.remove(executorId)
+                } else {
+                  waitingList(executorId) = workerLastHeartbeat
+                }
+                executorLastSeen.remove(executorId)
+              }
+            case None =>
+              for (executorId <- buf) {
+                val lastSeenMs = executorLastSeen.get(executorId).get
+                killExecutor(executorId, now - lastSeenMs)
+                executorLastSeen.remove(executorId)
+                waitingList.remove(executorId)
+              }
```

Comment:
Looks like the `if` condition should be `!sc.conf.get(HEARTBEAT_RECEIVER_CHECK_WORKER_LAST_HEARTBEAT) && sc.schedulerBackend isInstanceOf StandaloneSchedulerBackend` ?

Reply:
Update the `if` condition [1] to avoid other scheduler backends entering the `else` condition.
 
[1] https://github.com/apache/spark/pull/37411/files#diff-d1e46909e7b48d24379f8c373bc20e15f518da2f47655533d95bc947808fdd8fR299

Reply:
This (and other cases in this class) would get impacted due to lack of scheduler backend when the heatbeat receiver is created - please do revisit them after fixing that issue.

Reply:
Discussed below.

Reply:
Because we discussed this topic below, close this.

Reply:
@Ngone51 Here is one of the discussions about the lack of scheduler backend.